core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/AvroUpsertMonitor.scala

```diff
@@ -194,55 +210,88 @@ class AvroUpsertMonitorActor(
       self ! None
   }
 
-  private def initAvroUpsert(jobId: String, ackId: String): Future[Unit] = {
+  private def getPetServiceAccountUserInfo(workspaceName: WorkspaceName, userEmail: RawlsUserEmail) = {
     for {
-      avroMetadataJson <- readMetadataObject(s"$jobId/metadata.json")
-      //ack the response after we load the json into memory. pro: don't have to worry about ack timeouts for long operations, con: if someone restarts rawls here the uspert is lost
-      petSAJson <- samDAO.getPetServiceAccountKeyForUser(avroMetadataJson.namespace, RawlsUserEmail(avroMetadataJson.userEmail))
+      petSAJson <- samDAO.getPetServiceAccountKeyForUser(workspaceName.namespace, userEmail)
       petUserInfo <- googleServicesDAO.getUserInfoUsingJson(petSAJson)
-      avroUpsertJson <- readUpsertObject(s"$jobId/upsert.json")
+    } yield {
+      petUserInfo
+    }
+  }
+
+  private def markImportAsDone(importId: UUID, workspaceName: WorkspaceName, userInfo: UserInfo) = {
+    for {
+      importStatus <- importServiceDAO.getImportStatus(importId, workspaceName, userInfo)
+      _ <- importStatus match {
+        case Some(ImportStatuses.Upserting) => publishMessageToUpdateImportStatus(importId, ImportStatuses.Done)
+        case None => publishMessageToUpdateImportStatus(importId, ImportStatuses.Error("Import status not found"))
+        case _ => publishMessageToUpdateImportStatus(importId, ImportStatuses.Error("Upsert finished, but import status was not correct."))
+      }
+    } yield ()
+  }
+
+  private def publishMessageToUpdateImportStatus(importId: UUID, importStatus: ImportStatus) = {
+    val now = DateTime.now
+    val requiredAttributes = Map("importId" -> importId.toString, "updateStatus" -> importStatus.toString, "updatedDate" -> now.toString)
+    val attributes = requiredAttributes ++ {importStatus match {
+      case error: ImportStatuses.Error => Map("errorMessage" -> error.message)
+      case _ => Map()
+    }}
+    val message = GooglePubSubDAO.MessageRequest("", attributes)
+    pubSubDao.publishMessages(importStatusPubSubTopic, Seq(message))
+  }
+
+
+  private def initUpsert(upsertFile: String, jobId: UUID, ackId: String, workspaceName: WorkspaceName, userInfo: UserInfo): Future[Unit] = {
+    for {
+      //ack the response after we load the json into memory. pro: don't have to worry about ack timeouts for long operations, con: if someone restarts rawls here the upsert is lost
+      upsertJson <- readUpsertObject(upsertFile)
```

Comment:
Not for this PR, but I did a little digging into streaming the json rather than downloading it.

According to [this StackOverflow](https://stackoverflow.com/a/50999067/2941784) it's possible to create a Java NIO `InputStream` from a bucket object. And [circe-fs2](https://github.com/circe/circe-fs2) claims to be able to create a stream from a JSON array-of-objects. There's probably a way to stitch these together, in which case you could grab a few and make a batch out of them.

Reply:
https://broadworkbench.atlassian.net/browse/AS-201